### PR TITLE
Various minor fixes

### LINKED ISSUE #108: Async exception handling implemented incorrectly
The current approach to safely handling expcetions thrown in async queries etc. is not sufficient and needs to be improved.

### LINKED ISSUE #160: Table order needs to be defined for import
The tables for each dataset type need to be imported in a specific order, but finding them through the attributes orders them alphabetically, causing the import to fail.

### PR DESCRIPTION
**Pull request by MikeStower**
Resolves #160 
Resolves #108

### CODE REVIEW

**Comment by MikeStower** [comment]
514005c91c59bc4eeada92cb1a7903149af1f092 resolves #161

**Comment by jbrider** [comment]
This is labelled minor fixes and you have changed the overall design structure?

**Comment by MikeStower** [comment]
In hindsight, that was a poor choice of words on my part. My thinking was that since I wasn't editing the code too much, just a couple of file locations / namespaces, meaning it wasn't a huge change. It sounds like a lot because I removed one of the projects, but that project only had 2-3 files that didn't really belong in a separate layer anyway.

**Comment by jbrider** [comment]
To save me looking through 44 changed files - can you explain what was moved between layers and why?

**Comment by MikeStower** [comment]
I was trying to improve the quality of error messages that popped up for the user, but I'd been a bit sloppy with my implementation of async exception handling. This is why failed imports could cause multiple message boxes in a row to pop up, as well as that 'object reference is null' exception that you saw regularly. 

The original code had multiple nested try-catch blocks across different layers. These layers used also had different approaches to sending MediatR queries, which necessitated different exception handling.

By moving the excel importer and apsimx exporter, I could use the same code to send queries for everything. See commit f8339c4e369aca5d57c438fedfdaf68731867e3a.

Then, I added a method to handle exceptions in async tasks, which could be used with this new code to catch everything at the same point. See commit 00b90c517aac918c08ec815bc8932403dc4ebaed.

In short - I standardised how queries are called, and I moved the exception handling on top of this new standard code, instead of it being haphazardly used all over the place.

**Comment by jbrider** [comment]
Standardising the exception is good - and necessary.
The infrastructure layer is there to keep the implementation details out of the gui - so we can implement a different gui (web) or use a different type of database or export to Apsim Classic instead of Apsim Nextgen. The actual import/export code should be in the Infrastructure layer - hidden behind interfaces. 
The interaction between the gui/infrastructure should be done using Entities. I'm not sure that the import/export entities need to be in the Domain layer - but they should be in the Infrastrucure layer.  A concrete example is that the Apsim export should be hidden behind an interface so we can implement either nextgen or classic - we have chosen nextgen in this instance. 
Handling errors shouldn't require lifting those implementation details into the gui layer - that is a failing of the error handling.

**Comment by MikeStower** [comment]
I agree with everything you said about the infrastructure layer, and at some point I plan on adding it back in. I realised as I was fixing this though, that the presentation layer had grown too much and a lot of its functionality does actually belong in an infrastructure layer.

However, going the other direction, moving that functionality from presentation to infrastructure was a much messier prospect, especially given the bugs / poor design choices present that required fixing, and I judged that I should prioritise fixing the error messages for Greg. I opted to temporarily remove the infrastructure layer so that I could simplify and fix the code first, since it would also make it easier to extract the appropriate logic when the time came to add the infrastructure back in.

Basically, I was just untangling everything so it can be put back together nicely later on.

**Comment by jbrider** [comment]
I think you have gone about this backwards.
If it doesn't need interaction with the gui then it should be pushed to a lower layer.
If the issue is too much integration with the gui, then the it needs to be separated and the non gui pushed down into inrfrastructure.
I should be able to change the gui to a web site and still be able to import data.

**Comment by MikeStower** [comment]
Sorry if I didn't make it clearer - that is absolutely the end goal I've been working towards while making these changes. Many of the models in Presentation will be moved into a new Infrastructure layer at some point. However, to that end, the code needed to be debugged and refactored first, which I found easier and faster to do if all the models were temporarily in the same layer.
